### PR TITLE
NEW: Task runner better UI

### PR DESCRIPTION
**Pull request by mfendeksilverstripe**
# Task runner better UI

Simple UI changes which improve usability of the task runner. Lightweight CSS (no build chain or JS required).

## Problem

* it takes time to find the dev task you want to run especially if you don't know the exact task name
* links which execute dev tasks are tiny, mis-click error can happen
* current UI is cluttered with too many dev tasks and the space is used inefficiently
* rendering component of the task runner is messy (combines backend and frontend logic)
* overriding frontend display requires subclassing

### Current UI
Screen Shot 2020-06-08 at 12 05 05 PM

## Solution

* move the task groups into separate tabs so the screen is less cluttered
* visually separate the dev tasks
* visually separate the action buttons and make them much more visible
* split the rendering into backend (populate data) and a template (render), this allows the template to be overridden without subclassing

### Proposed UI
Screen Shot 2020-06-08 at 1 16 11 PM

## Related issues

https://github.com/symbiote/silverstripe-queuedjobs/issues/302

## Dependencies

https://github.com/silverstripe/silverstripe-framework/pull/9540

### CODE REVIEW

File: .editorconfig
Comment (michalkleiner) [inline .editorconfig]: Just make it 4 spaces everywhere... ;-)

Comment (mfendeksilverstripe) [reply .editorconfig]: Done and done :)

Comment (michalkleiner) [reply .editorconfig]: I expect some pushback from the core team as it's not a part of the functional change, but tbh I never understood the need for 2 spaces for `.yml` and `.json` only.

**Comment by michalkleiner** [comment]
Very nice touch! Often these sparingly used areas of the CMS/admin interface don't get any updates/tweaks, so this is refreshing to see.

I'd suggest looking at it at the framework level though, if possible, where it lives without the queued jobs additions (run now/enqueue). Here it can subsequently alter the actions.

**Comment by mfendeksilverstripe** [comment]
Will have a look at the core task runner. Thanks for feedback @michalkleiner

**Comment by mfendeksilverstripe** [comment]
@michalkleiner Moved some of the code to the framework, please review.

File: .editorconfig
Comment (emteknetnz) [reply .editorconfig]: Just copy whatever is in https://github.com/silverstripe/silverstripe-framework/blob/4/.editorconfig

> the need for 2 spaces

I don't know either, though consistent standards mean we don't need to think or spend time debating stuff =)

Comment (mfendeksilverstripe) [reply .editorconfig]: @emteknetnz  Editor config updated from framework

**Review by sachajudd — CHANGES_REQUESTED** [review]
Review comments for framework can be found here but also apply to this PR: https://github.com/silverstripe/silverstripe-framework/pull/9540#pullrequestreview-441332266

**Comment by mfendeksilverstripe** [comment]
@sachajudd Pushed up new changes based on UI feedback, please review. This is how the UI looks like now:

## Large screen
Screen Shot 2020-07-17 at 11 37 53 AM

## Small screen
Screen Shot 2020-07-17 at 11 38 22 AM

File: client/styles/task-runner.css
Comment (sachajudd) [inline client/styles/task-runner.css]: What's this colour for? Should it be #43536d?

File: client/styles/task-runner.css
Comment (sachajudd) [inline client/styles/task-runner.css]: Looks a bit tight in the UI, try 15px

**Comment by sachajudd** [comment]
See comments on https://github.com/silverstripe/silverstripe-framework/pull/9540#issuecomment-659760122 which apply to this PR 🙂

File: client/styles/task-runner.css
Comment (mfendeksilverstripe) [reply client/styles/task-runner.css]: good catch, thanks :)

**Comment by mfendeksilverstripe** [comment]
New changes:

- border hover colour fixed
- more space for bottom border

**Comment by mfendeksilverstripe** [comment]
New changes:

- updates labels in other tabs.

This is now ready for a review.

**Review by sachajudd — APPROVED** [review]
Approved from a UX perspective 🙂

File: client/styles/task-runner.css
Comment (Cheddam) [inline client/styles/task-runner.css]: Love the technique - and I think you could take it a step further. At the moment you're grouping the tasks into separate sections, which can result in some inconsistent spacing between items at the split. Instead, you could group them back together, shift the modifier to the `.task__item` level, and use an adjusted selector to hide / show individual tasks:

```suggestion
.task__selector--immediate:checked ~ .task__panel .task__item--immediate,
.task__selector--universal:checked ~ .task__panel .task__item--universal,
.task__selector--queue-only:checked ~ .task__panel .task__item--queue-only {
    display: block;
}
```

This would also mean you could reduce duplication in the template, and instead have conditionals around the task buttons.

Of course, keeping the types split into separate sections might be desirable. If this is the case, I'd at least suggest adding a heading to each section to make this separation clearer.

File: templates/Symbiote/QueuedJobs/Controllers/QueuedTaskRunner.ss
Comment (Cheddam) [inline templates/Symbiote/QueuedJobs/Controllers/QueuedTaskRunner.ss:12]: I think the default tab should make all tasks visible, rather than only 'universal' tasks. That might mean having four tabs if you want to keep this filter.

File: client/styles/task-runner.css
Comment (Cheddam) [inline client/styles/task-runner.css]: I don't think these rules (along with the hardcoded widths on the `input`s) are necessary to align / size these elements. I get almost identical output (pushed to the right by a couple of pixels) by dropping both of these sets of selectors out, and changing the margin on `.task__selector` to `0`.

**Review by Cheddam — CHANGES_REQUESTED** [review]
Hey Mojmir! This PR, along with the core one, is a great step forward for the dev/tasks UI. I've suggested some fairly minor modifications.

I'm also pondering whether it'd be sensible to reduce the code duplication between the base TaskRunner and this one by splitting some of the functionality out into separate methods / config? For example, the stylesheets could be stored in configuration, and the logic for handling them could be split into a method on the base TaskRunner class. This might be overkill, but it might also serve to reduce the amount of double-handling required in future updates to this UI (for example, I'm keen to introduce proper categorisation of tasks later on, along with a search field.)

Let me know what you think.

File: client/styles/task-runner.css
Comment (mfendeksilverstripe) [reply client/styles/task-runner.css]: Great suggestion, thanks :)

File: client/styles/task-runner.css
Comment (mfendeksilverstripe) [reply client/styles/task-runner.css]: Nice catch, let's remove it :)

**Comment by mfendeksilverstripe** [comment]
@Cheddam Pushed up changes as requested, please review.

**Comment by mfendeksilverstripe** [comment]
@Cheddam Pushed up changes as requested, please review.

**Review by Cheddam — APPROVED** [review]
Excellent work, thanks for your contribution!

**Review by Cheddam — CHANGES_REQUESTED** [review]
Actually, one last minor tweak is required 😅 

Since we standardised the way the CSS is included in the framework PR, this PR now needs to bump the minimum framework requirement to `^4.7` (the next minor release). I'll tweak this now.

**Review by Cheddam — APPROVED** [review]
Okay, I've also rebased onto the `4` branch since this is a new feature, and fixed up Travis config. Once the tests go green we're good to merge 👍

**Pull request merged**
